[RFC] Allow passing lua functions and closures into vim functions.

tjdevries opened the pull request:
Thanks to the great work by @prabirshrestha at https://github.com/vim/vim/pull/6246 I was able to implement his plan for nvim's code base (with slight modifications using our APIs and some clean up work to get some associated required changes).

So far so good it seems. Not sure if any other major features are required to make this compat w/ the PR on Vim's side.

Closes: #12138 
Closes: #7607

tjdevries committed:
lua: Add ability to pass lua functions directly to vimL

prabirshrestha commented:
I added support for `__call` would be in vim would be good to port that code too.

tjdevries commented:
Ha, I just about to comment that it appears you have beat me to that feature :) I will also implement it for us then before I merge. I'm going camping for the weekend, so I'll pick this up some time next week.

btw, are there any other `vim.*` things you like in the Vim API that don't for nvim? Just wondering if we can add some (perhaps 99% equivalent) wrappers around them to get easy compat.

prabirshrestha commented:
`vim.eval` doesn't exist in neovim. need to use `vim.api.nvim_eval' instead. To me it seems to be the same but haven't looked at the internal so don't know the detail about neovim. Would be good if we can have `vim.eval`

My another big issues is using `+1` for arrays index in neovim. This ones always leads to bugs.  Can we align?

https://github.com/prabirshrestha/vim-lsp/pull/821/files
https://github.com/prabirshrestha/asyncomplete.vim/pull/144

tjdevries commented:
I can probably just do `vim.eval = vim.api.nvim_eval` and I'd expect that to be fine.

I'm not sure exactly what you mean with the +1 for arrays index in neovim. Are you talking about  the lua arrays starting at 1? It wasn't clear to me from the two links you had.

prabirshrestha commented:
`vim.eval = vim.api.nvim_eval` should be fine.

In vim it starts with `0` and neovim it starts with `1`. I'm not sure if this is a vim bug or neovim since both are right in one or the other. You could argue that aligning with lua would allow it to interop with other libraries while aligning with vim will allow you to interop with vim.

Also noticed that neovim doesnt have `vim.dict()` and `vim.list()`. Might be better to start a separate thread on this and trying to align. Would also be good if we can port the tests from vim.

prabirshrestha commented:
@tjdevries Enjoy your camping! Have fun.

jamessan commented on src/nvim/eval/userfunc.c:
```diff
@@ -162,6 +163,17 @@ static void register_closure(ufunc_T *fp)
+/// Get a name for a lambda.  Returned in static memory.
+char_u * get_lambda_name(void)
+{
+    static char_u   name[30];
```
This isn't thread safe.  I know right now we only have the main and tui threads, so there _shouldn't_ be a problem but it seems cleaner to just have the caller provide the buffer.

tjdevries replied:
I just ported this over from vim's code -- they added this in a later commit. I didn't patch the whole thing though.

tjdevries replied:
Oh, I think I see what you're saying. Instead of taking `void`, I should just pass in `char_u *` pointer and let it set there, so there's no problem with getting the wrong buffer or some other dumb race condition.

tjdevries replied:
Hmmm, I don't think I'll actually do this in this PR, since I've just ported the vim implementation for this if that's OK. Do we have some note to leave for not thread-safe stuff?

norcalli replied:
I also don't think this makes sense. It seems safe in its usage, but it also seems like a rather unnecessary change.

tjdevries commented:
@prabirshrestha Can you make an issue where we can track & discuss other Lua incompatibilities that would be nice to fix? Not sure I can guarantee we'll fix all of them, but we can try and fix some of them at least.

I'll work on this some more this week.

prabirshrestha commented:
@tjdevries added some of the incompatibilities at https://github.com/neovim/neovim/issues/12537.

vim patch has been merged.

tjdevries commented:
@prabirshrestha I think I need to wait to implement `__call` since it is more complicated for us since we don't use the userdata stuff the same way as vim. I also want some of the changes from the autocmd PR to do this. So instead I will remove references to `__call` from this PR at the moment and implement it later.

Thanks for making the issue.

tjdevries committed:
lua: Add ability to pass tables with __call

vim-patch:8.2.1054: not so easy to pass a lua function to Vim
vim-patch:8.2.1084: Lua: registering function has useless code

I think I have also opened up the possibility for people to use these
callbacks elsewhere, since I've added a new struct that we should be
able to use.

Also, this should allow us to determine what the state of a list is in
Lua or a dictionary in Lua, since we now can track the luaref as we go.

norcalli commented:
I still think making the _call part a separate PR makes sense so that in the future someone can easily discover the discussion around it (especially since the content of the PR doesn't match the title anymore). It makes it a little more difficult to approve/review just one part of it, which is the ability to pass lua functions to vim callbacks.

norcalli commented on src/nvim/eval/userfunc.c:
```diff
@@ -3435,3 +3452,29 @@ bool set_ref_in_func(char_u *name, ufunc_T *fp_in, int copyID)
+  STRCPY(fp->uf_name, name);
+  hash_add(&func_hashtab, UF2HIKEY(fp));
+
+  return name;
```
Considering that you're returning a pointer to static memory that may be overriden, I'm not sure this is a good idea anymore.

tjdevries commented:
Hmm, when I pushed my ammended commit, it dropped your comment @norcalli : This is What I was writing:

> Oh, duh, I had switched this at one point to fp->uf_name.... I'll push. Should be fixed now @norcalli

norcalli commented:
Pending tests, I'll approve this and make a follow up PR to clean up some things I didn't like from reviewing.

tjdevries commented:
Sounds good to me :) ty.

norcalli commented:
Test did fail though. unused variable lint. Shouldn't our compilation or something check those locally? https://builds.sr.ht/~jmk/job/252153#task-build-634

tjdevries commented:
Huh, it didn't fail on my lint for this PR?... weird. https://travis-ci.org/github/neovim/neovim/jobs/707068009

tjdevries commented:
Huh, it's used on this line: https://github.com/neovim/neovim/pull/12507/files#diff-619381246c51a32ff4cf5de3e7cd627cR1433

Maybe it's because it's in an assert? That's weird though. Looks like tests passed for travis though. Waiting on appveyor.

norcalli reviewed (approved):
So long as tests pass, this looks good for now. I have some very minor tweaks I want to send which can come in a follow up PR whenever.

tjdevries commented on src/nvim/lua/vim.lua:
```diff
@@ -272,6 +272,10 @@ vim.fn = setmetatable({}, {
   end
 })
 
+vim.funcref = function(viml_func_name)
```
Ah, it's vim-lua compat -- they have `vim.funcref`.

Borwe commented:
Can one give an example of how to use this? I am assuming it would work with vim.fn.termopen(), so anyway to get it to map on_exit to a lua function without using vimscript rappers?

tjdevries commented:
```lua
:echo map([1, 2, 3], luaeval("function(k, v) return string.format('lua:%s:%s', k, v) end"))
```

noib3 commented:
Hi, I'm not sure if this already fixed a problem I'm having. I'm trying to call a vimscript function that takes (the equivalent of) a Lua table as a parameter, where one of the keys -- `sinklist` -- is a funcref. I tried the following
```lua
vim.fn['fzf#run'](vim.fn['fzf#wrap']({
  sinklist = 'v:lua.open_ripgreps',
}))
```
where I have defined
```lua
_G.open_ripgreps = function(...)
  -- whatever
end
```
but unfortunately I get a `Vim(call):E718: Funcref required` error. Is there a way to do this without touching Vimscript?

fsouza commented:
@noib3 discourse is probably a better place to ask questions, but you're not - [ ] passing a funcref. You're passing a string.

To take advantage of this PR you'd simply pass the function:

```
vim.fn['fzf#run'](vim.fn['fzf#wrap']({
  sinklist = function(...) end,
}))
```

noib3 commented:
> discourse is probably a better place to ask questions, but you're not passing a funcref. You're passing a string.

@fsouza oops, sorry about that, I'll move this to discourse (PS: that didn't work).
